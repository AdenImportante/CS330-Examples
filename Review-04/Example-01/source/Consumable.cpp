#include "Consumable.h"

/**
 * 
 */
Consumable::Consumable()
    : Item( std::string(), true )
{
    this->uses = 0;
}

/**
 * 
 */
Consumable::Consumable( const Consumable &src )
    : Item( src.name, true ),
    effect( src.effect )
{
    this->uses = src.uses;
}

/**
 *
 */
void Consumable::display( std::ostream &outs ) const{
    outs << "  Nme: " << Item::name << "\n"
         << "  Eft: " << effect     << "\n"
         << "  Use: " << uses       << "\n";
}

/**
 * 
 */
void Consumable::read( std::istream& ins ){
    ins >> Item::name
        >> effect
        >> uses;
}

Item* Consumable::clone() const{
    return new Consumable( *this );
}