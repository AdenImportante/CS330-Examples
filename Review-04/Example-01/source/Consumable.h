#ifndef CONSUMABLE_H_INCLUDED
#define CONSUMABLE_H_INCLUDED

#include "Item.h"

/**
 * This class represents one Consumable Item--as found in most video games. This includes
 * food.
 *
 * Consumable Items must be stackable. All Constructors must initialize Item::stackable
 * to true.
 */
class Consumable : public Item{
    private:        

    protected:
        std::string effect; ///< effect recieved by using the Item  
        int uses;           ///< number of times this Item can be used   

    public:
        /**
         * Default to a Comsumable Item with an empty name
         */
        Consumable();

        /**
         * Copy Constructor
         */
        Consumable( const Consumable &src );

        /**
         * Retrieve effect
         */
        std::string getEffect() const;

        /**
         * Set effect
         */
        void setEffect( std::string effect );

        /**
         * Retrieve number of uses
         */
        int getNumberOfUses() const;

        /**
         * Set number of uses
         */
        void setNumberOfUses( int u );

        /**
         * Print the Consumable Item
         */
        virtual void display( std::ostream &outs ) const;

        /**
         * Read Consumable Item attributes from an input stream
         */
        virtual void read( std::istream& ins );

        /**
         * Clone--i.e., copy--this Consumable Item
         */
        virtual Item* clone() const;
};

/**
 * 
 */
inline std::string Consumable::getEffect() const{
    return this->effect;
}

/**
 * 
 */
inline void Consumable::setEffect( std::string effect ){
    this->effect = effect;
}

/**
 * 
 */
inline int Consumable::getNumberOfUses() const{
    return this->uses;
}

/**
 * 
 */
inline void Consumable::setNumberOfUses( int u ){
    this->uses = u;
}

#endif