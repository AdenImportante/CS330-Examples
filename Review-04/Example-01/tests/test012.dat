#! /bin/sh

# Test Consumable::read and Consumable::display
# formatting must match

cat <<EOF > test012.cpp

#include <iostream>
#include <iomanip>
#include <sstream>

#include "Consumable.h"

using std::cout;
using std::istringstream;

int main(){
    Item *consumable = new Consumable(); 

    istringstream ins( "Luck-II-Potion Luck+50% 1" );

    consumable->read( ins );    
    consumable->display( cout );  

    return 0;
}
EOF

g++ -c test012.cpp 
g++ -o test012 test012.o Consumable.o Item.o

./test012