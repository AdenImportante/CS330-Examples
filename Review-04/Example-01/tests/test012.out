  Nme: Luck-II-Potion
  Eft: Luck+50%
  Use: 1
